### DISCUSSION TITLE
SPARQLStore returns Exception when query from fuseki

### DISCUSSION THREAD

**Discussion by cz23773** [RDFLib/rdflib#1632] [Q&A] [answered]
**Issue**

rdflib.plugins.stores SPARQLStore returns exception when query the fuseki running in a local docker, the code never be changed and this issue starts to happen recently.

**Versions**
- rdflib 5.0.0 
- Python 3.8
- Fuseki  3.17.0

**code**
The fuseki_endpoint will be redirected by nginx server.
```
import rdflib
from rdflib.plugins.stores import sparqlstore

fuseki_endpoint = "http://fuseki:3030/graph_name"
g = sparqlstore.SPARQLStore(f"{url}/query", "jsonld")
result = g.query("SELECT * WHERE { ?s ?p ?o . } LIMIT 1")

```

**Error**

Traceback (most recent call last):

>   File "/app/./core/helpers.py", line 171, in load_graph
>     result = g.query("SELECT * WHERE { ?s ?p ?o . } LIMIT 1")
>   File "/usr/local/lib/python3.8/site-packages/rdflib/plugins/stores/sparqlstore.py", line 186, in query
>     return self._query(query,
>   File "/usr/local/lib/python3.8/site-packages/rdflib/plugins/stores/sparqlstore.py", line 150, in _query
>     return super(SPARQLStore, self).query(*args, **kwargs)
>   File "/usr/local/lib/python3.8/site-packages/rdflib/plugins/stores/sparqlconnector.py", line 98, in query
>     return Result.parse(BytesIO(res.content), content_type=res.headers['Content-type'])
>   File "/usr/local/lib/python3.8/site-packages/rdflib/query.py", line 211, in parse
>     parser = plugin.get(plugin_key, ResultParser)()
>   File "/usr/local/lib/python3.8/site-packages/rdflib/plugin.py", line 107, in get
>     return p.getClass()
>   File "/usr/local/lib/python3.8/site-packages/rdflib/plugin.py", line 69, in getClass
>     module = __import__(self.module_path, globals(), locals(), [""])
>   File "/usr/local/lib/python3.8/site-packages/rdflib/plugins/sparql/__init__.py", line 33, in <module>
>     from . import parser
>   File "/usr/local/lib/python3.8/site-packages/rdflib/plugins/sparql/parser.py", line 184, in <module>
>     Param('prefix', PN_PREFIX)) + Suppress(':').leaveWhitespace()
>   File "/usr/local/lib/python3.8/site-packages/rdflib/plugins/sparql/parserutils.py", line 114, in __init__
>     self.name = name
> AttributeError: can't set attribute

**Comment by ghost** [comment, marked as answer]
The error suggests that your issue is caused by backwards-incompatible changes in pyparsing 3.0 which you can resolve by reverting to pyparsing 2.4.7
```bash
$ pip uninstall pyparsing
$ pip install pyparsing==2.4.7
```
or, if that's a problem, applying the following patch:
```diff
diff --git a/rdflib/plugins/sparql/parserutils.py b/rdflib/plugins/sparql/parserutils.py
index 29804eea..24ad6d9b 100644
--- a/rdflib/plugins/sparql/parserutils.py
+++ b/rdflib/plugins/sparql/parserutils.py
@@ -111,9 +111,9 @@ class Param(TokenConverter):
     """
 
     def __init__(self, name, expr, isList=False):
-        self.name = name
         self.isList = isList
         TokenConverter.__init__(self, expr)
+        self.setName(name)
         self.addParseAction(self.postParse2)
 
     def postParse2(self, tokenList):
@@ -215,7 +215,7 @@ class Comp(TokenConverter):
     def __init__(self, name, expr):
         self.expr = expr
         TokenConverter.__init__(self, expr)
-        self.name = name
+        self.setName(name)
         self.evalfn = None
 
     def postParse(self, instring, loc, tokenList):

```
Although the specific pyparsing version wasn't/isn't pinned in requirements.txt, it probably should have been and that aspect of RDFLib distributions will be now be under consideration.

I can't comment on the timescale for a bugfix release of RDFLib 5.0.0, so I'll leave this issue open in case other RDFLib 5 users encounter the same problem.

**Reply by CaptSolo** [reply]
A similar issue appears when running the `Graph.query()` method in the version of RDFLib that is available on the ChatGPT Code Interpreter: https://x.com/CaptSolo/status/1814400073574265284

Users do not have control over the versions of libraries installed there. The software versions installed are:

- rdflib: 6.0.0
- pyparsing: 3.1.2

Could you suggest how to work around this issue (e.g. using monkeypatching)?

Let me know if I should open a new issue about this.

![image](https://github.com/user-attachments/assets/9ca73134-3fe3-4146-b529-0ae9b76f4697)
